### PR TITLE
Feature/388/add option sort tree map by audit size

### PR DESCRIPTION
**Pull request by RomanenkoVladimir**
# Sort TreeMap by Size Metric or Alphabetically

closes feature request #388 

## Description

In order to find big packages in the TreeMap
We should provide an option to sort the TreeMap by the current size metric for the auditor
whereas currently he has to hover over each package individually to find large ones.

Given the map codecharta.cc.json
When I sort by size
Then the tree structure should be the same
and inside visualization/app/codecharta/the sort order should be ui, core, codecharta.component.scss etc.

This implementation splits the requested feature into a button to allow to reverse the order and an option dialog. The dialog would enable to add additional sorting options in the future.

## Definition of Done

A task/pull request will not be considered to be complete until all these items can be checked off.

- [x] **All** requirements mentioned in the issue are implemented
- [x] Does match the Code of Conduct and the Contribution file
- [x] Task has its own **GitHub issue** (something it is solving)
  - [x] Issue number is **included in the commit messages** for traceability
- [x] **Update the README.md** with any changes/additions made
- [x] **Update the CHANGELOG.md** with any changes/additions made
- [x] **Enough test coverage to ensure that uncovered changes do not break functionality**
- [x] **All tests pass**
- [x] **Descriptive pull request text**, answering:
  - What problem/issue are you fixing?
  - What does this PR implement and how?
- [x] **Assign your PR to someone** for a code review
  - This person _will be contacted **first**_ if a bug is introduced into `master`
- [x] **Manual testing** did not fail

### CODE REVIEW

File: visualization/app/codeCharta/ui/mapTreeView/mapTreeView.component.spec.ts [hunk 1]
```diff
@@ -39,11 +52,81 @@ describe("MapTreeViewController", () => {
...
+			SortingOrderAscendingService.subscribe = jest.fn()
+
+			rebuildController()
+
+			expect(SortingOrderAscendingService.subscribe).toHaveBeenCalledWith($rootScope, mapTreeViewController)
+		})
+	})
+
+	describe("applySort", () => {
+		it("should return the children array sorted by unary", () => {
```
Comment (NearW) [inline visualization/app/codeCharta/ui/mapTreeView/mapTreeView.component.spec.ts]: also here in the test suite: use const, use empty lines between arrange, act, assert blocks

File: visualization/app/codeCharta/ui/mapTreeView/mapTreeView.component.spec.ts [hunk 2]
```diff
@@ -39,11 +52,81 @@ describe("MapTreeViewController", () => {
...
+		it("should subscribe to SortingOrderAscendingService", () => {
+			SortingOrderAscendingService.subscribe = jest.fn()
+
+			rebuildController()
+
+			expect(SortingOrderAscendingService.subscribe).toHaveBeenCalledWith($rootScope, mapTreeViewController)
+		})
+	})
+
+	describe("applySort", () => {
```
Comment (NearW) [inline visualization/app/codeCharta/ui/mapTreeView/mapTreeView.component.spec.ts]: after applySort is private, I'd just test it implicit with applySortOrderChange()

File: visualization/app/codeCharta/ui/mapTreeView/mapTreeView.component.spec.ts [hunk 3]
```diff
@@ -3,15 +3,26 @@ import { MapTreeViewController } from "./mapTreeView.component"
...
+import { SortingDialogOptionService } from "../../state/store/dynamicSettings/sortingDialogOption/sortingDialogOption.service"
+import { SortingOrderAscendingService } from "../../state/store/appSettings/sortingOrderAscending/sortingOrderAscending.service"
 
 describe("MapTreeViewController", () => {
 	let mapTreeViewController: MapTreeViewController
 	let $rootScope: IRootScopeService
 	let $timeout: ITimeoutService
+	let storeService = getService<StoreService>("storeService")
 	let map: CodeMapNode
+	let validSetupNode: CodeMapNode
```
Comment (NearW) [inline visualization/app/codeCharta/ui/mapTreeView/mapTreeView.component.spec.ts]:
```suggestion
	let mapWithMultipleFolders: CodeMapNode
```

File: visualization/app/codeCharta/util/dataMocks.ts [hunk 4]
```diff
@@ -1,17 +1,18 @@
 import {
```
Comment (NearW) [inline visualization/app/codeCharta/util/dataMocks.ts]: After this PR is cleaned up, we should remove the `VALID` prefix from every variable here, since there are no invalid cases given. Even if we will declare some, we can prefix them with INVALID. It's just not needed to do it for valid stuff.

**Comment by NearW** [comment]
@RomanenkoVladimir Please update the changelog as well here. :)

File: visualization/app/codeCharta/ui/sortingOptionDialog/sortingOptionDialog.component.ts [hunk 5]
```diff
@@ -0,0 +1,40 @@
+import "./sortingOptionDialog.component.scss"
```
[the remark this answers is not in the document]
Comment (alschmut) [reply visualization/app/codeCharta/ui/sortingOptionDialog/sortingOptionDialog.component.ts]: As its not a pop-up dialog like the others (which use the $mdDialog component from angular), it's definitely a separate component. I would therefore remove the name dialog here, as it could become confusing. How about just `sortingOption.component` @RomanenkoVladimir, @NearW? :)

@RomanenkoVladimir unfortunately renaming a component is not as easy as creating one 😄 some manual refactoring is necessary.. if you have questions, always feel free to ask 🎉

File: visualization/app/codeCharta/ui/mapTreeView/mapTreeView.component.ts [hunk 6]
```diff
@@ -1,27 +1,86 @@
...
 			return
 		}
-
 		this._viewModel.rootNode = map
 		this.synchronizeAngularTwoWayBinding()
+
+		this.storeService.dispatch(setSortingDialogOption(this.storeService.getState().dynamicSettings.sortingDialogOption))
+		if (this.storeService.getState().appSettings.sortingOrderAscending) {
+			this.storeService.dispatch(setSortingOrderAscending(this.storeService.getState().appSettings.sortingOrderAscending))
+		}
```
[the remark this answers is not in the document]
Comment (alschmut) [reply visualization/app/codeCharta/ui/mapTreeView/mapTreeView.component.ts]: @NearW `onFilesSelectionChanged` might not work, as the treeView displays the nodes of the renderMap (which is stored inside the codeMap.preRenderService), which is not available yet, when the event is sent.

If it is worth to increase performance here we could do following:
```
let collectNextRenderMap = false

onBlacklistChanged(blacklist: BlacklistItem[]) {
  collectNextRenderMap = true
}

onFilesSelectionChanged(...) {
  collectNextRenderMap = true
}

onRenderMapChanged(map: CodeMapNode) {
  if (collectNextRenderMap) {
    this._viewModel.rootNode = map
    collectNextRenderMap = false
    // ...
  }
}
```

I will leave the decision up to you.

File: visualization/app/codeCharta/ui/matchingFilesCounter/matchingFilesCounter.component.html [hunk 7]
```diff
@@ -1,3 +1,7 @@
-<span class="box-rounded node-count">{{ $ctrl._viewModel.fileCount }}<i class="fa fa-file"></i></span>
+<span class="box-rounded node-count ">{{ $ctrl._viewModel.fileCount }}<i class="fa fa-file"></i></span>
 <span class="box-rounded node-count">{{ $ctrl._viewModel.flattenCount }}<i class="fa fa-eye-slash"></i></span>
 <span class="box-rounded node-count">{{ $ctrl._viewModel.excludeCount }}<i class="fa fa-ban"></i></span>
+<div class="sort-component-styling">
+	<sorting-button-component id="id-sort-component-button"></sorting-button-component>
+	<sorting-option-dialog-component id="id-sort-component-dialog"></sorting-option-dialog-component>
```
[the remark this answers is not in the document]
Comment (RomanenkoVladimir) [reply visualization/app/codeCharta/ui/matchingFilesCounter/matchingFilesCounter.component.html]: the ids are necessary, or the default styling of md-button or md-select-value would take precedence, an alternative would be using !important, but such actions are discoraged

File: visualization/app/codeCharta/ui/sortingOptionDialog/sortingOptionDialog.component.spec.ts [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+	})
+
+	describe("onSortingDialogOptionChanged", () => {
+		it("should update the sortingOptionDialog to Childnodes", () => {
+			let sortingDialogOption = SortingOption.Childnodes
+			sortingOptionDialogController.onSortingDialogOptionChanged(sortingDialogOption)
+			expect(sortingOptionDialogController["_viewModel"].sortingOption).toEqual(SortingOption.Childnodes)
+		})
+		it("should update the sortingOptionDialog to Name", () => {
+			let sortingDialogOption = SortingOption.Name
```
[the remark this answers is not in the document]
Comment (RomanenkoVladimir) [reply visualization/app/codeCharta/ui/sortingOptionDialog/sortingOptionDialog.component.spec.ts]: I thought that this would improve readability, but maybe you are right removing it will get rid of otherwise irrelevant lines of code

File: visualization/app/codeCharta/codeCharta.model.ts [hunk 9]
```diff
@@ -66,6 +66,11 @@ export enum NodeType {
 	FOLDER = "Folder"
 }
 
+export enum SortingOption {
+	Name = "Name",
+	Childnodes = "Childnodes"
```
[the remark this answers is not in the document]
Comment (RomanenkoVladimir) [reply visualization/app/codeCharta/codeCharta.model.ts]: I would prefer another string instead of "Number of Files", as the string is rather long, making a distinction between the sorting dialog and the matching-file-counter rather difficult
I will try to come up with one, otherwise proposals are really welcome

File: visualization/app/codeCharta/ui/mapTreeView/mapTreeView.component.ts [hunk 10]
```diff
@@ -1,27 +1,86 @@
...
 	/* @ngInject */
-	constructor(private $rootScope: IRootScopeService, private $timeout: ITimeoutService) {
+	constructor(private $rootScope: IRootScopeService, private $timeout: ITimeoutService, private storeService: StoreService) {
 		CodeMapPreRenderService.subscribe(this.$rootScope, this)
+		SortingDialogOptionService.subscribe(this.$rootScope, this)
+		SortingOrderAscendingService.subscribe(this.$rootScope, this)
+	}
+
+	public onSortingDialogOptionChanged(sortingDialogOption: SortingOption) {
+		if (sortingDialogOption == SortingOption.Childnodes) {
```
[the remark this answers is not in the document]
Comment (RomanenkoVladimir) [reply visualization/app/codeCharta/ui/mapTreeView/mapTreeView.component.ts]: good point
can we enable this tslint rule, so it will get checked automatically upon commit? 
(Rule: triple-equals)

File: visualization/app/codeCharta/util/dataMocks.ts (hunk 4 above)
Comment (RomanenkoVladimir) [reply visualization/app/codeCharta/util/dataMocks.ts]: I think this should be done as a separate Merge Request, not really part of the feature.
Or am I in the wrong here?

File: visualization/app/codeCharta/ui/mapTreeView/mapTreeView.component.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (RomanenkoVladimir) [reply visualization/app/codeCharta/ui/mapTreeView/mapTreeView.component.ts]: Could you elaborate on the change?
As of now mapTreeView.component does not subscribe to BlacklistService at all, this is handled by mapTreeView.level.component
Do you propose a restructuring here?

**Comment by RomanenkoVladimir** [comment]
I applied most of the requested changes, the ones not applied are 

- removing ids in css: not possible here
- subscribe to BlacklistService and listen to onBlacklistChanged: I am not sure what needs to be done, could someone provide an explanation?
- remove VALID tags: not part of the feature in my opinion

All tests pass and a bug connected to the sortingButton not working correctly upon reloading the map has been resolved.

File: visualization/app/codeCharta/util/dataMocks.ts (hunk 4 above)
Comment (NearW) [reply visualization/app/codeCharta/util/dataMocks.ts]: Agreed

File: visualization/app/codeCharta/codeCharta.model.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (NearW) [reply visualization/app/codeCharta/codeCharta.model.ts]: but isn't the matching-file-counter and the sorting dialog based on the same metric? 
I don't really see a reason to make a distinction here. Can you give me an example?

File: visualization/app/codeCharta/ui/matchingFilesCounter/matchingFilesCounter.component.html [hunk 11]
```diff
@@ -1,3 +1,7 @@
-<span class="box-rounded node-count">{{ $ctrl._viewModel.fileCount }}<i class="fa fa-file"></i></span>
-<span class="box-rounded node-count">{{ $ctrl._viewModel.flattenCount }}<i class="fa fa-eye-slash"></i></span>
-<span class="box-rounded node-count">{{ $ctrl._viewModel.excludeCount }}<i class="fa fa-ban"></i></span>
+<span class="box-rounded node-count ">{{ $ctrl._viewModel.fileCount }}<i class="fa fa-file"></i></span>
+<span class="box-rounded node-count ">{{ $ctrl._viewModel.flattenCount }}<i class="fa fa-eye-slash"></i></span>
+<span class="box-rounded node-count ">{{ $ctrl._viewModel.excludeCount }}<i class="fa fa-ban"></i></span>
```
Comment (NearW) [inline visualization/app/codeCharta/ui/matchingFilesCounter/matchingFilesCounter.component.html]: empty space at the end of the class name is still here.

File: visualization/app/codeCharta/ui/matchingFilesCounter/matchingFilesCounter.component.html [hunk 12]
```diff
@@ -1,3 +1,7 @@
...
+<div id="sort-component">
+	<sorting-button-component id="id-sort-component-button"></sorting-button-component>
```
Comment (NearW) [inline visualization/app/codeCharta/ui/matchingFilesCounter/matchingFilesCounter.component.html]: I meant to change the id from `id-sort-component-button` to `sort-component-button`

File: visualization/app/codeCharta/codeCharta.model.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (RomanenkoVladimir) [reply visualization/app/codeCharta/codeCharta.model.ts]: I mean in regards to the ui, the button get pretty close to the matching-file-counter icons
My bad on the wording here

File: visualization/app/codeCharta/ui/matchingFilesCounter/matchingFilesCounter.component.html (hunk 11 above)
Comment (RomanenkoVladimir) [reply visualization/app/codeCharta/ui/matchingFilesCounter/matchingFilesCounter.component.html]: i misunderstood the request and inserted empty spaces instead, should be fixed

File: visualization/app/codeCharta/ui/mapTreeView/mapTreeView.component.ts [hunk 13]
```diff
@@ -1,27 +1,86 @@
...
 
 	/* @ngInject */
-	constructor(private $rootScope: IRootScopeService, private $timeout: ITimeoutService) {
+	constructor(private $rootScope: IRootScopeService, private $timeout: ITimeoutService, private storeService: StoreService) {
 		CodeMapPreRenderService.subscribe(this.$rootScope, this)
+		SortingOptionService.subscribe(this.$rootScope, this)
+		SortingOrderAscendingService.subscribe(this.$rootScope, this)
+	}
+
+	public onSortingOptionChanged(sortingOption: SortingOption) {
```
Comment (NearW) [inline visualization/app/codeCharta/ui/mapTreeView/mapTreeView.component.ts]:
```
public onSortingOptionChanged(sortingOption: SortingOption) {
		if (sortingOption === SortingOption.NUMBER_OF_FILES) {
			this._viewModel.rootNode = this.applySortOrderChange(
				this._viewModel.rootNode,
				(a, b) => b.attributes["unary"] - a.attributes["unary"],
				false
			)
		} else {
this._viewModel.rootNode = this.applySortOrderChange(this._viewModel.rootNode, (a, b) => (b.name > a.name ? 0 : 1), false)
                 }	this.storeService.dispatch(setSortingOrderAscending(this.storeService.getState().appSettings.sortingOrderAscending))
	}
```

what about this solution?

Comment (RomanenkoVladimir) [reply visualization/app/codeCharta/ui/mapTreeView/mapTreeView.component.ts]: will desync the button if you load a map without changes
I will commit another fix now, that should resolve the button issue

**Comment by NearW** [comment]
@RomanenkoVladimir update changelog and add overflow for the sortOption
